modules/statistics/AnalyticsAdapter.js

```diff
@@ -1,130 +1,344 @@
+ *
+ *    action,
+ *    actionSubject,
+ *    actionSubjectId,
+ *    attributes,
+ *    categories,
+ *    containerId,
+ *    containerType,
+ *    name,
+ *    objectId,
+ *    objectType,
+ *    source,
+ *    tags
+ * }
+ *
+ * The 'type' is one of 'operational', 'page', 'track' or 'ui', and some of the
+ * other properties are considered required according to the type.
+ *
+ * For events with type 'page', the required properties are: name.
+ *
+ * For events with type 'operational' and 'ui', the required properties are:
+ * action, actionSubject, source
+ *
+ * For events with type 'page', the required properties are:
+ * action, actionSubject, source, containerType, containerId, objectType,
+ * objectId
  */
-class AnalyticsAbstract {
+class AnalyticsAdapter {
     /**
-     *
+     * Creates new AnalyticsAdapter instance.
      */
-    sendEvent() {} // eslint-disable-line no-empty-function
-}
+    constructor() {
+        /**
+         * Whether this AnalyticsAdapter has been disposed or not.
+         * @type {boolean}
+         */
+        this.disposed = false;
+
+        /**
+         * The set of handlers to which events will be sent.
+         * @type {Set<any>}
+         */
+        this.analyticsHandlers = new Set();
+
+        /**
+         * The cache of events which are not sent yet. The cache is enabled
+         * while this field is truthy, and disabled otherwise.
+         * @type {Array}
+         */
+        this.cache = [];
+
+        /**
+         * Map of properties that will be added to every event. Note that the
+         * keys will be prefixed with "permanent.".
+         */
+        this.permanentProperties = {};
+
+        this.addPermanentProperties({
+            callstatsname: Settings.callStatsUserName,
+            userAgent: navigator.userAgent,
+            browserName: RTCBrowserType.getBrowserName()
+        });
+    }
 
-/**
- * Handler that caches all the events.
- * @extends AnalyticsAbstract
- */
-class CacheAnalytics extends AnalyticsAbstract {
     /**
-     *
+     * Dispose analytics. Clears all handlers.
      */
-    constructor() {
-        super();
+    dispose() {
+        this.setAnalyticsHandlers([]);
+        this.disposed = true;
+    }
+
+    /**
+     * Sets the handlers that are going to be used to send analytics. Sends any
+     * cached events.
+     * @param {Array} handlers the handlers
+     */
+    setAnalyticsHandlers(handlers) {
+        if (this.disposed) {
+            return;
+        }
+
+        this.analyticsHandlers = new Set(handlers);
 
-        // some events may happen before init or implementation script download
-        // in this case we accumulate them in this array and send them on init
-        this.eventCache = [];
+        // Note that we disable the cache even if the set of handlers is empty.
+        const cache = this.cache;
+
+        this.cache = null;
+        if (cache) {
+            cache.forEach(event => this._sendEvent(event));
+        }
```

Comment:
Is there a guarantee or even a desire to make sure all analytics handlers for jitsi-meet get assigned at once so they get all the events? Like if an event is cached, and only one handler gets set, then that handler gets the event, but then a later handler that gets set wouldn't get the event. That's a pretty edge-casey situation and now that I've written it out doesn't seem like something worth handling.

Reply:
I'm assuming that once the user of the library calls setAnalyticsHandlers(), it is their responsibility to handle events. If they wish to handle the case that you describe, they should implement their own cache :)
But seriously: I don't want to keep the cache enabled just because of this. It's just meant to serve for event sent early on before all the initialization in the client has run